### ISSUE TITLE
Build Failure on MacOSX 10.15 - nanoflann

### ISSUE THREAD

**Issue by KrisBecker** [DOI-USGS/ISIS3#4742]
**ISIS version(s) affected**: 6.0.0+
ISIS source code cloned on 2022-01-10

**Description**  
After updating my ISISDATA and ISISTESTDATA areas, I attempted a build of ISIS on my Mac configuration (same as in #4656). A compile error was encounter on cnet2dem which is likely related to nanoflann:

```
(ErosDevE) zion2[1115]: ninja install
[1076/3600] Building CXX object objects/CMakeFiles/cnet2dem_app.dir/control/apps/cnet2dem/main.cpp.o
FAILED: objects/CMakeFiles/cnet2dem_app.dir/control/apps/cnet2dem/main.cpp.o 
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++ -DBOOST_ALL_NO_LIB -DBOOST_ATOMIC_DYN_LINK -DBOOST_CHRONO_DYN_LINK -DBOOST_CONTAINER_DYN_LINK -DBOOST_CONTEXT_DYN_LINK -DBOOST_DATE_TIME_DYN_LINK -DBOOST_FILESYSTEM_DYN_LINK -DBOOST_IOSTREAMS_DYN_LINK -DBOOST_LOG_DYN_LINK -DBOOST_LOG_SETUP_DYN_LINK -DBOOST_MATH_C99F_DYN_LINK -DBOOST_MATH_C99L_DYN_LINK -DBOOST_MATH_C99_DYN_LINK -DBOOST_MATH_TR1F_DYN_LINK -DBOOST_MATH_TR1L_DYN_LINK -DBOOST_MATH_TR1_DYN_LINK -DBOOST_PRG_EXEC_MONITOR_DYN_LINK -DBOOST_PROGRAM_OPTIONS_DYN_LINK -DBOOST_RANDOM_DYN_LINK -DBOOST_REGEX_DYN_LINK -DBOOST_SERIALIZATION_DYN_LINK -DBOOST_SYSTEM_DYN_LINK -DBOOST_TEST_EXEC_MONITOR_DYN_LINK -DBOOST_THREAD_DYN_LINK -DBOOST_TIMER_DYN_LINK -DBOOST_UNIT_TEST_FRAMEWORK_DYN_LINK -DBOOST_WAVE_DYN_LINK -DBOOST_WSERIALIZATION_DYN_LINK -DISISBUILDDIR=\"/Users/kbecker/ISIS/Isis4Dev/ErosPDART/ISIS3/build\" -DISISROOT=\"/Users/kbecker/ISIS/Isis4Dev/ErosPDART/ISIS3/isis\" -DJSON_DIAGNOSTICS=0 -DJSON_USE_IMPLICIT_CONVERSIONS=1 -DQT_CONCURRENT_LIB -DQT_CORE_LIB -DQT_GUI_LIB -DQT_MULTIMEDIAWIDGETS_LIB -DQT_MULTIMEDIA_LIB -DQT_NETWORK_LIB -DQT_NO_DEBUG -DQT_OPENGL_LIB -DQT_PRINTSUPPORT_LIB -DQT_QML_LIB -DQT_QUICK_LIB -DQT_SCRIPTTOOLS_LIB -DQT_SCRIPT_LIB -DQT_SQL_LIB -DQT_SVG_LIB -DQT_TESTCASE_BUILDDIR=\"/Users/kbecker/ISIS/Isis4Dev/ErosPDART/ISIS3/build\" -DQT_TESTLIB_LIB -DQT_WEBCHANNEL_LIB -DQT_WIDGETS_LIB -DQT_XMLPATTERNS_LIB -DQT_XML_LIB -I/Users/kbecker/ISIS/Isis4Dev/ErosPDART/ISIS3/build/inc -isystem /Users/kbecker/ISIS/Isis4Dev/ErosPDART/ISIS3/isis/../gtest/googletest -isystem /Users/kbecker/ISIS/Isis4Dev/ErosPDART/ISIS3/isis/../gtest/googletest/include -isystem /Users/kbecker/ISIS/Isis4Dev/ErosPDART/ISIS3/isis/../gtest/googlemock -isystem /Users/kbecker/ISIS/Isis4Dev/ErosPDART/ISIS3/isis/../gtest/googlemock/include -isystem /Users/kbecker/miniconda3/envs/ErosDevE/include/ale -isystem /Users/kbecker/miniconda3/envs/ErosDevE/include -isystem /Users/kbecker/miniconda3/envs/ErosDevE/include/bullet -isystem /Users/kbecker/miniconda3/envs/ErosDevE -isystem /usr/include/X11 -isystem /Users/kbecker/miniconda3/envs/ErosDevE/include/csm -isystem /Users/kbecker/miniconda3/envs/ErosDevE/include/cspice -isystem /Users/kbecker/miniconda3/envs/ErosDevE/include/eigen3 -isystem /Users/kbecker/miniconda3/envs/ErosDevE/include/embree2 -isystem /Users/kbecker/miniconda3/envs/ErosDevE/include/jama -isystem /Users/kbecker/miniconda3/envs/ErosDevE/include/nlohmann -isystem /Users/kbecker/miniconda3/envs/ErosDevE/include/opencv4/opencv2 -isystem /Users/kbecker/miniconda3/envs/ErosDevE/include/opencv4 -iframework /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/System/Library/Frameworks -isystem /Users/kbecker/miniconda3/envs/ErosDevE/include/pcl-1.11/pcl -isystem /Users/kbecker/miniconda3/envs/ErosDevE/include/pcl-1.11 -isystem /Users/kbecker/miniconda3/envs/ErosDevE/include/tnt -isystem /Users/kbecker/miniconda3/envs/ErosDevE/include/qt -isystem /Users/kbecker/miniconda3/envs/ErosDevE/include/qt/QtConcurrent -isystem /Users/kbecker/miniconda3/envs/ErosDevE/include/qt/QtCore -isystem /Users/kbecker/miniconda3/envs/ErosDevE/./mkspecs/macx-clang -isystem /Users/kbecker/miniconda3/envs/ErosDevE/include/qt/QtGui -isystem /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/System/Library/Frameworks/OpenGL.framework/Headers -isystem /Users/kbecker/miniconda3/envs/ErosDevE/include/qt/QtMultimediaWidgets -isystem /Users/kbecker/miniconda3/envs/ErosDevE/include/qt/QtMultimedia -isystem /Users/kbecker/miniconda3/envs/ErosDevE/include/qt/QtNetwork -isystem /Users/kbecker/miniconda3/envs/ErosDevE/include/qt/QtWidgets -isystem /Users/kbecker/miniconda3/envs/ErosDevE/include/qt/QtOpenGL -isystem /Users/kbecker/miniconda3/envs/ErosDevE/include/qt/QtPrintSupport -isystem /Users/kbecker/miniconda3/envs/ErosDevE/include/qt/QtQml -isystem /Users/kbecker/miniconda3/envs/ErosDevE/include/qt/QtQuick -isystem /Users/kbecker/miniconda3/envs/ErosDevE/include/qt/QtScriptTools -isystem /Users/kbecker/miniconda3/envs/ErosDevE/include/qt/QtScript -isystem /Users/kbecker/miniconda3/envs/ErosDevE/include/qt/QtSql -isystem /Users/kbecker/miniconda3/envs/ErosDevE/include/qt/QtSvg -isystem /Users/kbecker/miniconda3/envs/ErosDevE/include/qt/QtTest -isystem /Users/kbecker/miniconda3/envs/ErosDevE/include/qt/QtWebChannel -isystem /Users/kbecker/miniconda3/envs/ErosDevE/include/qt/QtXmlPatterns -isystem /Users/kbecker/miniconda3/envs/ErosDevE/include/qt/QtXml -Wall -fPIC -std=c++14 -DISIS_LITTLE_ENDIAN=1 -Wno-unused-parameter -Wno-overloaded-virtual -Wno-strict-aliasing -DUSE_UNSTABLE_GEOS_CPP_API=1 -Wno-strict-overflow -DENABLEJP2K=OFF -Wall -fPIC -std=c++14 -DISIS_LITTLE_ENDIAN=1 -Wno-unused-parameter -Wno-overloaded-virtual -Wno-strict-aliasing -DUSE_UNSTABLE_GEOS_CPP_API=1 -Wno-strict-overflow -DENABLEJP2K=OFF -O3 -DNDEBUG -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk -fPIC -std=gnu++14 -MD -MT objects/CMakeFiles/cnet2dem_app.dir/control/apps/cnet2dem/main.cpp.o -MF objects/CMakeFiles/cnet2dem_app.dir/control/apps/cnet2dem/main.cpp.o.d -o objects/CMakeFiles/cnet2dem_app.dir/control/apps/cnet2dem/main.cpp.o -c /Users/kbecker/ISIS/Isis4Dev/ErosPDART/ISIS3/isis/src/control/apps/cnet2dem/main.cpp
In file included from /Users/kbecker/ISIS/Isis4Dev/ErosPDART/ISIS3/isis/src/control/apps/cnet2dem/main.cpp:32:
/Users/kbecker/ISIS/Isis4Dev/ErosPDART/ISIS3/isis/src/control/apps/cnet2dem/PointCloudTree.h:107:67: error: non-const lvalue reference to type 'vector<pair<unsigned int, [...]>>' cannot bind to a value of unrelated type 'vector<pair<size_t, [...]>>'
        int n = m_kd_index.radiusSearch(point.array(), radius_sq, matches,
                                                                  ^~~~~~~
/Users/kbecker/ISIS/Isis4Dev/ErosPDART/ISIS3/isis/src/control/apps/cnet2dem/main.cpp:340:31: note: in instantiation of member function 'Isis::PointCloudTree<Isis::ControlPointCloudPt, Dist3Dto2D<Isis::ControlPointCloudPt> >::radius_query' requested here
            results = cloud_t.radius_query(cpt, search_radius_sq);
                              ^
/Users/kbecker/miniconda3/envs/ErosDevE/include/nanoflann.hpp:1458:61: note: passing argument to parameter 'IndicesDists' here
        std::vector<std::pair<AccessorType, DistanceType>>& IndicesDists,
                                                            ^
In file included from /Users/kbecker/ISIS/Isis4Dev/ErosPDART/ISIS3/isis/src/control/apps/cnet2dem/main.cpp:32:
/Users/kbecker/ISIS/Isis4Dev/ErosPDART/ISIS3/isis/src/control/apps/cnet2dem/PointCloudTree.h:117:56: error: cannot initialize a parameter of type 'unsigned int *' with an rvalue of type 'unsigned long *'
        m_kd_index.knnSearch(point.array(), neighbors, indices.data(),
                                                       ^~~~~~~~~~~~~~
/Users/kbecker/ISIS/Isis4Dev/ErosPDART/ISIS3/isis/src/control/apps/cnet2dem/main.cpp:342:33: note: in instantiation of member function 'Isis::PointCloudTree<Isis::ControlPointCloudPt, Dist3Dto2D<Isis::ControlPointCloudPt> >::neighbor_query' requested here
              results = cloud_t.neighbor_query(cpt, neighbors);
                                ^
/Users/kbecker/miniconda3/envs/ErosDevE/include/nanoflann.hpp:1430:23: note: passing argument to parameter 'out_indices' here
        AccessorType* out_indices, DistanceType* out_distances_sq,
                      ^
2 errors generated.
[1085/3600] Building CXX object objects/CMakeFiles/rolo2isis_app.dir/rolo/apps/rolo2isis/main.cpp.o
ninja: build stopped: subcommand failed.
``` 

It appears this is due to a very new release of **[nanoflann](https://anaconda.org/conda-forge/nanoflann)** - ~5 days ago.

**How to reproduce**  
Follow build procedures as outlined by [instructions](https://github.com/USGS-Astrogeology/ISIS3/wiki/Developing-ISIS3-with-cmake#getting-started-with-github).

**Possible Solution**  
Pin nanoflann < 1.4.0 or fix several usage occurrences in ISIS code base to run using nanoflann 1.4.0+.

**Additional context**  
The previous conda environment used several months ago (in #4656) shows an earlier version of nanoflann when compared to the new dev environment.

Old:
```
nanoflann                 1.3.2                h694c41f_0    conda-forge
```

New:
```
nanoflann                 1.4.0                h694c41f_0    conda-forge
```

NOTE: After I pinned nanoflann as mentioned above, the following warning is the only remaining artifact in the build:

```
(ErosDevEa) zion2[1135]: ninja install
[2514/3600] Building CXX object objects/CMakeFiles/isis.dir/qisis/objs/TrackTool/TrackTool.cpp.o
/Users/kbecker/ISIS/Isis4Dev/ErosPDART/ISIS3/isis/src/qisis/objs/TrackTool/TrackTool.cpp:289:3: warning: control may reach end of non-void function [-Wreturn-type]
  }
  ^
1 warning generated.
````

**Comment by KrisBecker** [comment]
Well its unfortunate that this fix is buried in what appears to be an unrelated issue. 

Wouldn't a separate PR allow the fix to be expedited rather than wait for #4732 to be closed?

**Comment by jlaura** [comment]
I rolled it into #4732 because it was a build failure in our CI and was expected to close quickly. ~~For a faster fix, PRs are always welcome.~~

Apologies, I missed the PR! @jessemapel  We could have merged that and then rebased the Jenkins PR. That would have been cleaner for @KrisBecker :smile:

**Comment by jlaura** [comment]
This was fixed by #4732. Please re-open if the issue persists.
